metricbeat/cmd/rollup.go

```diff
@@ -0,0 +1,101 @@
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package cmd
+
+import (
+	"fmt"
+
+	"os"
+
+	"github.com/spf13/cobra"
+
+	"github.com/elastic/beats/libbeat/cmd/instance"
+	"github.com/elastic/beats/libbeat/fields/rollup"
+	"github.com/elastic/beats/libbeat/template"
+)
+
+func GenRollupConfigCmd(name string) *cobra.Command {
+	genRollupConfigCmd := &cobra.Command{
+		Use:   "rollup",
+		Short: "Export rollup config for module/metricset to stdout",
+		Run: func(cmd *cobra.Command, args []string) {
+
+			// Skipping index prefix and beat version as not needed
+			b, err := instance.NewBeat(name, "", "")
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "Error creating beat: %s\n", err)
+				os.Exit(1)
+			}
+
+			module, _ := cmd.Flags().GetString("module")
+			metricSet, _ := cmd.Flags().GetString("metricset")
+			if module == "" || metricSet == "" {
+				fmt.Fprintf(os.Stderr, "Module and metricset params have to be set.")
+				os.Exit(1)
+			}
+
+			// Load fields from Beat
+			f, err := template.LoadYamlByte(b.Fields)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "Error loading fields: %s\n", err)
+				os.Exit(1)
+			}
+			nodePath := fmt.Sprintf("%s.%s", module, metricSet)
+			fields := f.GetNode(nodePath)
+
+			if fields == nil {
+				fmt.Fprint(os.Stderr, "No fields found for module %s and metricset %s\n", module, metricSet)
+				os.Exit(1)
+			}
+
+			// Load all the additional params
+			indexPattern, _ := cmd.Flags().GetString("index_pattern")
+			rollupIndex, _ := cmd.Flags().GetString("rollup_index")
+			cron, _ := cmd.Flags().GetString("cron")
+			pageSize, _ := cmd.Flags().GetString("page_size")
+			interval, _ := cmd.Flags().GetString("interval")
+			delay, _ := cmd.Flags().GetString("delay")
+
+			processor := rollup.NewProcessor()
+			err = processor.Process(fields, nodePath)
+			if err != nil {
+				fmt.Fprint(os.Stderr, "Error processing fields: %s", err)
+				os.Exit(1)
+			}
+
+			// Note: No validation happens if the strings used are actually valid
+			// TODO: What if no terms are specified? Should we return error?
+			_, err = os.Stdout.WriteString(processor.Generate(indexPattern, rollupIndex, cron, pageSize, interval, delay).StringToPrint() + "\n")
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "Error writing rollup job: %+v", err)
+				os.Exit(1)
+			}
+		},
+	}
+
+	genRollupConfigCmd.Flags().String("module", "", "Module to create rollup job for")
+	genRollupConfigCmd.Flags().String("metricset", "", "Metricset to create rollup job for")
+	genRollupConfigCmd.Flags().String("index_pattern", "metricbeat-*", "Index pattern to roll up on")
+	genRollupConfigCmd.Flags().String("rollup_index", "rollup-metricbeat", "Rollup index")
```

Comment:
@polyfractal Any recommendation on these defaults? I could make all the fields below required but I would prefer to have some defaults as otherwise it's by default a lot of configuration for the user.

Reply:
Some thoughts:

- I'd use a non-calendar interval (e.g. `60m` instead of `1h`) since they are a lot more flexible.  `1h` would limit them to other calendar units which are singular (`1h`, `1d`, `1w`, etc).  We [have a PR up](https://github.com/elastic/elasticsearch/pull/32052) to make this clear in Rollup, as well as converting all the docs over to fixed units
- If the interval is an hour, there's probably no need to set the cron to every 30s (it'll trigger, see there's nothing to do and go back to sleep).  Maybe every 10 or 30 minutes or something similar?
- `delay` is debatable.  It delays when we process a bucket of time, to allow for out-of-order data to show up.  If that's not an issue, I wouldn't even specify a delay.  Or set it to something close to the interval, like an hour.
- I'd go with a bigger `page_size`, maybe 1000?  This is the number of buckets we request from the composite agg, so a small page size == may small searches